### PR TITLE
feat: add cpu, memory and disk usage on Podman machines

### LINKED ISSUE #3063: Add resource usage for Podman machines
See https://github.com/containers/podman-desktop/issues/3063#issue-1780981835

### PR DESCRIPTION
**Pull request by jeffmaury**
Fixes #3063

### What does this PR do?

Add CPU, memory and disk usage (%) for Podman machines

### Screenshot/screencast of this PR

res

### What issues does this PR fix or reference?

Fixes #3063

### How to test this PR?

Go to the resource pages and display Podman machines

### CODE REVIEW

File: extensions/podman/src/extension.ts
Comment (benoitf) [inline extensions/podman/src/extension.ts]: would it be better to use the REST API instead ? so instead of executing a CLI and parsing JSON we would directly get JSON formatted response ?

Comment (jeffmaury) [reply extensions/podman/src/extension.ts]: Yeah the reason is that the context is different (machine vs extensionAPI but will look if I can find a mapping between machine name and engineId

Comment (benoitf) [reply extensions/podman/src/extension.ts]: yes parsing JSON output of a CLI would always lead to more errors than reusing the current connection (so no new process and we have direct JSON output) 👍

**Review by benoitf — APPROVED** [review]
would be better using REST API as here it'll probably introduce technical debt

but could be a follow-up

**Comment by benoitf** [comment]
@mairin could you confirm if the layout is 'OK' ?

image


I'm not sure if we need to have the full display with percentage number. I'm finding hard to read the values mixing both number and percentages with same height and after each value.

Do we need like small lines from green to orange / red depending on the percentage or other ideas 

I think there was a mockup with some metrics at some point

**Comment by mairin** [comment]
This is how the mockups display this data:

Screenshot from 2023-07-05 13-12-39

So the mockups have # CPU, VM memory allocation in GiB, VM max disk limit in GiB. 

@jeffmaury it looks like you have more data, here is what I am seeing:
- **# CPUs** (_same as in mockup_)
- CPU usage %
- disk size (is this the same as disk limit?)
- disk usage %
- **memory allocation** (_same as in mockup_)
- memory usage %

So I think the main difference here is that you are showing the allocations and the % usage. My question for the % usage - is that a live % usage, or is it an average?

**Comment by deboer-tim** [comment]
IMHO putting both the configured value and current usage % as labels all in a row makes this hard to read, too busy, and wider than it needs to be. I think we could do something with little usage bars or dials, colouring on low resources (e.g. red when >90% used), or tooltips to see exact values.

If I see that (e.g.) only 39% of disk and 59% of memory is being used and decide to shrink my machine to save space, you currently wouldn't see this information on the edit page either. It would be nice if there was a corresponding visualization on the slider (e.g. colouring or marker) so that you could see current use and don't need to go back and forth.

**Comment by mairin** [comment]
Screenshot from 2023-07-05 14-34-59

Just a quick idea, this is kinda taking inspiration from the PatternFly way of showing utilization charts... (eg https://www.patternfly.org/v4/charts/donut-utilization-chart ) - and you could provide the exact % or value on hover as shown on the RAM one on the right?

**Comment by benoitf** [comment]
@mairin I like it, important information is quickly visible

**Comment by jeffmaury** [comment]
@mairin Introduced a new UI component feel free to comment and tune

image

**Comment by lstocchi** [comment]
> Introduced a new UI component feel free to comment and tune

Is this ready to be reviewed?

**Comment by jeffmaury** [comment]
> > Introduced a new UI component feel free to comment and tune
> 
> Is this ready to be reviewed?

Yes as @mairin is ok for the new component

**Review by cdrage — COMMENTED** [review]
Code LGTM. But that font.. it's really small and I'm barely able to read it. Any way we can increase it? And instead of using names like "Memory" we can just say "MEM" so the larger font fits within the circle?

What do you think @mairin

**Comment by cdrage** [comment]
251246170-37fe06e9-47ad-45de-b5cf-691536196176
 
 Have it look more like the above

**Comment by deboer-tim** [comment]
Mairin's sample has the units smaller, which looks great but also makes it harder to center the text. IMHO a simpler change like replacing the text elements with what's below would clean it up enough that it's perfectly acceptable to merge for now, and we can improve the look separately.

```
  <text x="{size / 2}" y="37%" text-anchor="middle" class="text-xs stroke-gray-800">{title}</text>
  <text x="{size / 2}" y="51%" dominant-baseline="central" text-anchor="middle" class="text-sm stroke-gray-500"
    >{value}</text>
```

Screenshot 2023-07-11 at 3 21 15 PM

File: packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte
Comment (lstocchi) [inline packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte]: I face this error
```
filesize.esm.js:86 Uncaught (in promise) TypeError: Invalid number
    at filesize (filesize.esm.js:86:9)
    at Array.create_if_block_3 (PreferencesResourcesRendering.svelte:397:35)
    at create_each_block_3 (PreferencesResourcesRendering.svelte:393:111)
    at create_if_block_1 (PreferencesResourcesRendering.svelte:383:25)
    at Object.update [as p] (PreferencesResourcesRendering.svelte:381:74)
    at Object.update [as p] (PreferencesResourcesRendering.svelte:361:17)
    at Object.update [as p] (PreferencesResourcesRendering.svelte:303:11)
    at update_slot_base (index.mjs:100:14)
    at Object.update [as p] (SettingsPage.svelte:8:43)
    at update (index.mjs:1347:36)
```

**Review by lstocchi — CHANGES_REQUESTED** [review]

**Review by lstocchi — COMMENTED** [review]
It doesn't refresh the resources used when starting a machine. I have to force reload to actually see them.
In the console i see `Can't get machine podman-machine-default resource usage error Error: no running provider for the matching engine`

no_update_consumption

File: packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte
Comment (lstocchi) [reply packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte]: Fixed, i'm not able to reproduce it

**Comment by cdrage** [comment]
> It doesn't refresh the resources used when starting a machine. I have to force reload to actually see them. In the console i see `Can't get machine podman-machine-default resource usage error Error: no running provider for the matching engine`
> 
> no_update_consumption

Code looks great and thank you for updating the UI changes @jeffmaury 

I am getting the same issue as @lstocchi when testing out the PR.

**Comment by cdrage** [comment]
I think after a rebase we can get a review done / merge it in. I'm no longer getting the issue with the podman machine not appearing for showing the resources.

**Review by lstocchi — COMMENTED** [review]
Still not working for me. I'm facing the same error as before.
The `engine.api` is undefined in my case. The machine does work correctly though, i can see it in the terminal.
So it's like it appears to be running but the api is not set yet before the info function is executed.
If you need any other test/info let me know

**Review by lstocchi — COMMENTED** [review]
Tested, works fine for me now. I only have a question for my LGTM
I see that the value in digits is not actually following the donut visualization value.

image
By looking at the image above it looks like the machine is using 16GB of memory but the Donut tells me that it is only 5% of the total, which is wrong. So is the Donut wrong or the machine is not using 16Gb of memory?
Shouldn't the number in the Donut be the same as the one used to color it?

The other things i would update (these can be done in following PRs) are:
1) instead of seeing 0cpu, 0disk size, 0 memory when a machine get started, there should be a loading icon or something else that gives an idea that the values are updating (question for @mairin )
2) i find the donut a bit too large, but maybe it's just personal taste

**Comment by jeffmaury** [comment]
> Tested, works fine for me now. I only have a question for my LGTM I see that the value in digits is not actually following the donut visualization value.
> 
> image By looking at the image above it looks like the machine is using 16GB of memory but the Donut tells me that it is only 5% of the total, which is wrong. So is the Donut wrong or the machine is not using 16Gb of memory? Shouldn't the number in the Donut be the same as the one used to color it?
> 
> The other things i would update (these can be done in following PRs) are:
> 
>     1. instead of seeing 0cpu, 0disk size, 0 memory when a machine get started, there should be a loading icon or something else that gives an idea that the values are updating (question for @mairin )
> 
>     2. i find the donut a bit too large, but maybe it's just personal taste

The digits represent the capacity of the VM (ie max) where the donut represents the current use

**Comment by lstocchi** [comment]
> The digits represent the capacity of the VM (ie max) where the donut represents the current use

Ok, if that was made on purpose, everything works for me. GJ 🚀

**Review by lstocchi — APPROVED** [review]
LGTM

File: packages/main/src/plugin/container-registry.ts
Comment (benoitf) [inline packages/main/src/plugin/container-registry.ts]: hello, I think here we should prefix info to say it's a LibPodInfo because usually when there is no prefix it means it's coming from the compat API

File: extensions/podman/package.json
Comment (benoitf) [inline extensions/podman/package.json]: the format should be `memoryUsage`

File: extensions/podman/package.json
Comment (benoitf) [inline extensions/podman/package.json]: format is `cpuUsage`

File: extensions/podman/package.json
Comment (benoitf) [inline extensions/podman/package.json]: format is diskSizeUsage

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts]: we can't import types externally in that file

it needs to be self-contained

File: packages/extension-api/src/extension-api.d.ts
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts]: Here I think we're breaking a pattern as usually we were providing compat API

here it's a Podman API that is returned so I don't know if Info object should return info from the Compat API

**Review by benoitf — CHANGES_REQUESTED** [review]

File: packages/main/src/plugin/container-registry.ts
Comment (benoitf) [reply packages/main/src/plugin/container-registry.ts]: how does it work when you're using Docker engine ?

**Review by cdrage — APPROVED** [review]
There's a minor rebase, but otherwise this code LGTM / still works as before. 

Resources show up / update accordingly.

**Comment by benoitf** [comment]
on my side the rendering is not the same than the mockup

mockup = 
image


what I got (especially for the default disk size)

image

image

so it seems that when unit is with 3 digits on the left and 2 after it doesn't work

I see also 'Disk Size' vs 'Disk' and 'MEM' vs 'Memory' I don't know if units should be rounded or find something to render it better

cc @mairin

**Review by benoitf — APPROVED** [review]
code is 👍 all the issues reported are fixed
good job 🎉 

I just added remarks about the UI that feels not so nice as it's the default rendering with the default settings

**Comment by deboer-tim** [comment]
FWIW, I was planning to use Donut in another place (container statistics) and started fixing some additional things I found, e.g.:
- thinner grey circle to match the design
- use palette colors
- minor viewbox/scaling issues
- fix end limits (e.g. if the client incorrectly requests 125% the circle stops at 100%)
- wasn't changing colors if the value is live and crosses thresholds (e.g. green to red)
- added tooltip

The UI issues listed above are mostly not these, so I guess I'd suggest we merge and split the issues: I can release improvements to Donut, and we track another issue for use of it in this context, e.g. 'Disk Size', 'Memory'.

Separating the number vs units would allow Donut to consistently round values (and maybe use different font size like the design one day too).

**Pull request merged**
